modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java

```diff
@@ -92,40 +92,90 @@ protected String computeId(Context ctx, XAnnotatedObject xObject, Element elemen
         return id;
     }
 
-    @Override
-    @SuppressWarnings("unchecked")
-    protected <T> T doRegister(Context ctx, XAnnotatedObject xObject, Element element, String extensionId) {
-        String id = computeId(ctx, xObject, element);
+    protected boolean shouldRemove(Context ctx, XAnnotatedObject xObject, Element element, String extensionId,
+            String id) {
         XAnnotatedMember remove = xObject.getRemove();
         if (remove != null && Boolean.TRUE.equals(remove.getValue(ctx, element))) {
-            contributions.remove(id);
-            return null;
+            return true;
         }
-        Object contrib;
+        return false;
+    }
+
+    protected boolean shouldMerge(Context ctx, XAnnotatedObject xObject, Element element, String extensionId, String id,
+            Object existing) {
         XAnnotatedMember merge = xObject.getMerge();
         if (merge != null && Boolean.TRUE.equals(merge.getValue(ctx, element))) {
-            Object contribution = contributions.get(id);
-            if (contribution != null && xObject.getCompatWarnOnMerge() && !merge.hasValue(ctx, element)) {
+            if (existing != null && xObject.getCompatWarnOnMerge() && !merge.hasValue(ctx, element)
+                    && !onlyHandlesEnablement(ctx, xObject, element)) {
                 log.warn(
                         "The contribution with id '{}' on extension '{}' has been implicitly merged: "
                                 + "the compatibility mechanism on its descriptor class '{}' detected it, "
                                 + "and the attribute merge=\"true\" should be added to this definition.",
-                        id, extensionId, contribution.getClass().getName());
+                        id, extensionId, existing.getClass().getName());
             }
-            contrib = xObject.newInstance(ctx, element, contribution);
-        } else {
-            contrib = xObject.newInstance(ctx, element);
+            return true;
         }
-        contributions.put(id, contrib);
+        return false;
+    }
+
+    protected boolean onlyHandlesEnablement(Context ctx, XAnnotatedObject xObject, Element element) {
+        // checks no children
+        if (element.getChildNodes().getLength() > 0) {
+            return false;
+        }
+        // checks no attribute other than id and enablement
+        int nbAttr = element.getAttributes().getLength();
+        if (nbAttr > 2) {
+            return false;
+        }
+        XAnnotatedMember enable = xObject.getEnable();
+        return (enable != null && enable.hasValue(ctx, element));
+    }
```

Comment:
Can't it be put in the Abstract? Maybe other methods are concerned too.

Reply:
Actually i need to fix this: the implementation should differ between map and single (single registry should not expect an id attribute).
Other similar methods also have that constraint (need to extract the contribution and maybe expose the item id), but I'll have another look to see if some of this can be factored out.

Reply:
I factored out the shouldEnable and shouldRemove methods in the end, kept the merge methods apart (as the compat warn message needs the contribution id anyway, and the "onlyHandlesEnablement" method should be specific as explained above.